src/pages/write-smart-contracts/billboard-tutorial.md

```diff
@@ -0,0 +1,281 @@
+an empty test file to the project:
+
+```sh
+clarinet new billboard-clarity && cd billboard-clarity
+clarinet contract new billboard
+```
+
+These commands create the necessary project structure and contracts for completing this tutorial. Remember that at
+any point during this tutorial you can use `clarinet check` to check the validity of your Clarity syntax.
+
+## Step 2: create message storage
+
+Open the `contracts/billboard.clar` file in a text editor or IDE. For this tutorial, you'll use the boilerplate comments
+to structure your contract for easy readability.
+
+In this step, you'll add a variable to the contract that stores the billboard message, and define a getter function to
+read the value of the variable.
+
+Under the `data maps and vars` comment, define the `billboard-message` variable. Remember that you must define the type of
+the variable, in this case `string-utf8` to support emojis and extended characters. You must also define the
+maximum length of the variable, for this tutorial use the value `500` to allow for a longer message. You must also
+define the initial value for the variable.
+
+```clarity
+;; data maps and vars
+(define-data-var billboard-message (string-utf8 500) u"Hello world!")
+```
+
+You also should define a read-only getter function returns the value of the `billboard-message` variable.
+
+```clarity
+;; public functions
+(define-read-only (get-message)
+    (var-get billboard-message))
+```
+
+These are the required methods for storing and accessing the message on the billboard.
+
+## Step 3: define set message function
+
+Define a method to set the billboard message. Under the public functions, define a `set-message` function. This public
+function takes a `string-utf8` with a max length of `500` as the only argument. Note that the type of the argument
+matches the type of the `billboard-message` variable. Clarity's type checking ensures that an invalid input to the
+function doesn't execute.
+
+```clarity
+(define-public (set-message (message (string-utf8 500)))
+    (var-set billboard-message message)
+)
+```
+
+The contract is now capable of updating the `billboard-message`.
+
+## Step 4: transfer STX to set message
+
+In this step, you'll modify the `set-message` function to add a cost in STX tokens, that increments by a set amount each
+time the message updates.
+
+First, you should define a variable to track the price of updating the billboard. This value is in micro-STX. Under the
+`data maps and vars` heading, add a new variable `price` with type `uint` and an initial value of `u100`. The initial
+cost to update the billboard is 100 micro-STX or 0.0001 STX.
+
+```clarity
+(define-data-var price uint u100)
+```
+
+You also should define a read-only getter function returns the value of the `price` variable.
+
+```clarity
+(define-read-only (get-price)
+    (var-get price)
+)
+```
+
+It's a best practice to define codes to a descriptive constant for Clarity smart contracts. This makes the code easier
```

Comment:
should we explain why error codes are required?

Reply:
> This makes the code easier to understand for readers.

Are there any other reasons?

Reply:
I had these in mind:
- there are no error types in clarity, one convention became to use consts and give them error codes aligned with HTTP errors
- this makes errors reusable across methods

Reply:
> there are no error types in clarity, one convention became to use consts and give them error codes aligned with HTTP errors

Is this something the community does? The example contracts I've seen usually just use arbitrary numbers that count up from 0.

Reply:
ah, you're right - it's a mix from what I can tell from recent contracts: https://search-clarity.dev/